#**ANALISE DESCRITIVA** - Banco de Dados de avaliações *AWS*


# **1. Carregamento e Entendimento dos Dados**
Nessa etapa realizaremos o carregamento organizado de todas as tabelas do projeto no Python (Google Colab ou VS Code), além de compreender o que cada tabela representa, como se relacionam e qual o papel de cada uma no fluxo de um e-commerce. Também construiremos um dicionário de dados simples, descrevendo cada variável, seu tipo, significado e função no conjunto de informações.


## **1.1 Importação das Bibliotecas**
Aqui importamos as bibliotecas essenciais para manipulação, análise e visualização dos dados, como
Pandas, Numpy e Matplotlib.

In [ ]:
#importação das bibliotecas
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## **1.2 Carregamento dos dados**
Nessa etapa carregaremos todas as tabelas do projeto no Python utilizando a biblioteca Pandas. Utilize
caminhos corretos e específicos de suas pastas. Carregue as nove tabelas: pedidos, itens do pedido,
clientes, produtos, pagamentos, avaliações, vendedores, geolocalização e auxiliar.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
#importação das tabelas
import pandas as pd
#avaliações
df_avaliacoes = pd.read_csv('/content/drive/MyDrive/Banco_de_Dados_PII1_AWS/Banco_de_Dados_PII3_AWS/avaliacoes.csv')

#clientes
df_clientes = pd.read_csv('/content/drive/MyDrive/Banco_de_Dados_PII1_AWS/Banco_de_Dados_PII3_AWS/clientes.csv')

#geolocalizacao
df_geolocalizacao = pd.read_csv('/content/drive/MyDrive/Banco_de_Dados_PII1_AWS/Banco_de_Dados_PII3_AWS/geolocalizacao.csv')

#itens_pedidos
df_itens_pedidos = pd.read_csv('/content/drive/MyDrive/Banco_de_Dados_PII1_AWS/Banco_de_Dados_PII3_AWS/itens_pedidos.csv')

#pagamentos
df_pagamentos = pd.read_csv('/content/drive/MyDrive/Banco_de_Dados_PII1_AWS/Banco_de_Dados_PII3_AWS/pagamentos.csv')

#pedidos
df_pedidos = pd.read_csv('/content/drive/MyDrive/Banco_de_Dados_PII1_AWS/Banco_de_Dados_PII3_AWS/pedidos.csv')

#produtos
df_produtos = pd.read_csv('/content/drive/MyDrive/Banco_de_Dados_PII1_AWS/Banco_de_Dados_PII3_AWS/produtos.csv')

#tabela_auxiliar
df_tabela_auxiliar = pd.read_csv('/content/drive/MyDrive/Banco_de_Dados_PII1_AWS/Banco_de_Dados_PII3_AWS/tabela_auxiliar.csv')

#vendedores
df_vendedores = pd.read_csv('/content/drive/MyDrive/Banco_de_Dados_PII1_AWS/Banco_de_Dados_PII3_AWS/vendedores.csv')

In [ ]:

# Dicionario das tabelas do Banco de dados
tabelas = {
    'avaliacoes': df_avaliacoes,
    'clientes': df_clientes,
    'geolocalizacao': df_geolocalizacao,
    'itens_pedidos': df_itens_pedidos,
    'pagamentos': df_pagamentos,
    'pedidos': df_pedidos,
    'produtos': df_produtos,
    'tabela_auxiliar': df_tabela_auxiliar,
    'vendedores': df_vendedores
}

# Referenciando o tamanho de cada tabela no banco de dados
for nome, df in tabelas.items():
    print(f'{nome}: {df.shape[0]} linhas X {df.shape[1]} colunas \n{'-'*50}')

avaliacoes: 99224 linhas X 7 colunas 
--------------------------------------------------
clientes: 99441 linhas X 5 colunas 
--------------------------------------------------
geolocalizacao: 1000163 linhas X 5 colunas 
--------------------------------------------------
itens_pedidos: 112650 linhas X 7 colunas 
--------------------------------------------------
pagamentos: 103886 linhas X 5 colunas 
--------------------------------------------------
pedidos: 99441 linhas X 8 colunas 
--------------------------------------------------
produtos: 32951 linhas X 9 colunas 
--------------------------------------------------
tabela_auxiliar: 71 linhas X 2 colunas 
--------------------------------------------------
vendedores: 3095 linhas X 4 colunas 
--------------------------------------------------


In [ ]:
from pandas.core.dtypes.missing import isnull
def analise_df(nome, df):
  print(f"\n{'='*50}")
  print(f"\nTABELA: {nome}")
  print(f"\n{'='*50}")

  print("\nHEAD")
  display(df.head())

  print("\nTAIL")
  display(df.tail())

  print("\nINFO")
  df.info()

  print("\nDESCRIBE")
  display(df.describe())

  print("\nItens Nulos")
  print(df.isnull())

In [ ]:
for nome, df in tabelas.items():
    analise_df(nome, df)



TABELA: avaliacoes


HEAD


,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela I...,2018-03-01 00:00:00,2018-03-02 10:26:53



TAIL


,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
99219,574ed12dd733e5fa530cfd4bbf39d7c9,2a8c23fee101d4d5662fa670396eb8da,5,NaN,NaN,2018-07-07 00:00:00,2018-07-14 17:18:30
99220,f3897127253a9592a73be9bdfdf4ed7a,22ec9f0669f784db00fa86d035cf8602,5,NaN,NaN,2017-12-09 00:00:00,2017-12-11 20:06:42
99221,b3de70c89b1510c4cd3d0649fd302472,55d4004744368f5571d1f590031933e4,5,NaN,"Excelente mochila, entrega super rápida. Super...",2018-03-22 00:00:00,2018-03-23 09:10:43
99222,1adeb9d84d72fe4e337617733eb85149,7725825d039fc1f0ceb7635e3f7d9206,4,NaN,NaN,2018-07-01 00:00:00,2018-07-02 12:59:13
99223,efe49f1d6f951dd88b51e6ccd4cc548f,90531360ecb1eec2a1fbb265a0db0508,1,NaN,"meu produto chegou e ja tenho que devolver, po...",2017-07-03 00:00:00,2017-07-03 21:01:49



INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                99224 non-null  object
 1   order_id                 99224 non-null  object
 2   review_score             99224 non-null  int64 
 3   review_comment_title     11568 non-null  object
 4   review_comment_message   40977 non-null  object
 5   review_creation_date     99224 non-null  object
 6   review_answer_timestamp  99224 non-null  object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB

DESCRIBE


,review_score
count,99224.000000
mean,4.086421
std,1.347579
min,1.000000
25%,4.000000
50%,5.000000
75%,5.000000
max,5.000000



Itens Nulos
       review_id  order_id  review_score  review_comment_title  \
0          False     False         False                  True   
1          False     False         False                  True   
2          False     False         False                  True   
3          False     False         False                  True   
4          False     False         False                  True   
...          ...       ...           ...                   ...   
99219      False     False         False                  True   
99220      False     False         False                  True   
99221      False     False         False                  True   
99222      False     False         False                  True   
99223      False     False         False                  True   

       review_comment_message  review_creation_date  review_answer_timestamp  
0                        True                 False                    False  
1                        True       

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP



TAIL


,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
99436,17ddf5dd5d51696bb3d7c6291687be6f,1a29b476fee25c95fbafc67c5ac95cf8,3937,sao paulo,SP
99437,e7b71a9017aa05c9a7fd292d714858e8,d52a67c98be1cf6a5c84435bd38d095d,6764,taboao da serra,SP
99438,5e28dfe12db7fb50a4b2f691faecea5e,e9f50caf99f032f0bf3c55141f019d99,60115,fortaleza,CE
99439,56b18e2166679b8a959d72dd06da27f9,73c2643a0a458b49f58cea58833b192e,92120,canoas,RS
99440,274fa6071e5e17fe303b9748641082c8,84732c5050c01db9b23e19ba39899398,6703,cotia,SP



INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB

DESCRIBE


,customer_zip_code_prefix
count,99441.000000
mean,35137.474583
std,29797.938996
min,1003.000000
25%,11347.000000
50%,24416.000000
75%,58900.000000
max,99990.000000



Itens Nulos
       customer_id  customer_unique_id  customer_zip_code_prefix  \
0            False               False                     False   
1            False               False                     False   
2            False               False                     False   
3            False               False                     False   
4            False               False                     False   
...            ...                 ...                       ...   
99436        False               False                     False   
99437        False               False                     False   
99438        False               False                     False   
99439        False               False                     False   
99440        False               False                     False   

       customer_city  customer_state  
0              False           False  
1              False           False  
2              False           False  
3             

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP



TAIL


,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
1000158,99950,-28.068639,-52.010705,tapejara,RS
1000159,99900,-27.877125,-52.224882,getulio vargas,RS
1000160,99950,-28.071855,-52.014716,tapejara,RS
1000161,99980,-28.388932,-51.846871,david canabarro,RS
1000162,99950,-28.070104,-52.018658,tapejara,RS



INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  object 
 4   geolocation_state            1000163 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 38.2+ MB

DESCRIBE


,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng
count,1.000163e+06,1.000163e+06,1.000163e+06
mean,3.657417e+04,-2.117615e+01,-4.639054e+01
std,3.054934e+04,5.715866e+00,4.269748e+00
min,1.001000e+03,-3.660537e+01,-1.014668e+02
25%,1.107500e+04,-2.360355e+01,-4.857317e+01
50%,2.653000e+04,-2.291938e+01,-4.663788e+01
75%,6.350400e+04,-1.997962e+01,-4.376771e+01
max,9.999000e+04,4.506593e+01,1.211054e+02



Itens Nulos
         geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                              False            False            False   
1                              False            False            False   
2                              False            False            False   
3                              False            False            False   
4                              False            False            False   
...                              ...              ...              ...   
1000158                        False            False            False   
1000159                        False            False            False   
1000160                        False            False            False   
1000161                        False            False            False   
1000162                        False            False            False   

         geolocation_city  geolocation_state  
0                   False              False  
1   

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14



TAIL


,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
112645,fffc94f6ce00a00581880bf54a75a037,1,4aa6014eceb682077f9dc4bffebc05b0,b8bc237ba3788b23da09c0f1f3a3288c,2018-05-02 04:11:01,299.99,43.41
112646,fffcd46ef2263f404302a634eb57f7eb,1,32e07fd915822b0765e448c4dd74c828,f3c38ab652836d21de61fb8314b69182,2018-07-20 04:31:48,350.00,36.53
112647,fffce4705a9662cd70adb13d4a31832d,1,72a30483855e2eafc67aee5dc2560482,c3cfdc648177fdbbbb35635a37472c53,2017-10-30 17:14:25,99.90,16.95
112648,fffe18544ffabc95dfada21779c9644f,1,9c422a519119dcad7575db5af1ba540e,2b3e4a2a3ea8e01938cabda2a3e5cc79,2017-08-21 00:04:32,55.99,8.72
112649,fffe41c64501cc87c801fd61db3f6244,1,350688d9dc1e75ff97be326363655e01,f7ccf836d21b2fb1de37564105216cc1,2018-06-12 17:10:13,43.00,12.79



INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB

DESCRIBE


,order_item_id,price,freight_value
count,112650.000000,112650.000000,112650.000000
mean,1.197834,120.653739,19.990320
std,0.705124,183.633928,15.806405
min,1.000000,0.850000,0.000000
25%,1.000000,39.900000,13.080000
50%,1.000000,74.990000,16.260000
75%,1.000000,134.900000,21.150000
max,21.000000,6735.000000,409.680000



Itens Nulos
        order_id  order_item_id  product_id  seller_id  shipping_limit_date  \
0          False          False       False      False                False   
1          False          False       False      False                False   
2          False          False       False      False                False   
3          False          False       False      False                False   
4          False          False       False      False                False   
...          ...            ...         ...        ...                  ...   
112645     False          False       False      False                False   
112646     False          False       False      False                False   
112647     False          False       False      False                False   
112648     False          False       False      False                False   
112649     False          False       False      False                False   

        price  freight_value  
0      

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45



TAIL


,order_id,payment_sequential,payment_type,payment_installments,payment_value
103881,0406037ad97740d563a178ecc7a2075c,1,boleto,1,363.31
103882,7b905861d7c825891d6347454ea7863f,1,credit_card,2,96.80
103883,32609bbb3dd69b3c066a6860554a77bf,1,credit_card,1,47.77
103884,b8b61059626efa996a60be9bb9320e10,1,credit_card,5,369.54
103885,28bbae6599b09d39ca406b747b6632b1,1,boleto,1,191.58



INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB

DESCRIBE


,payment_sequential,payment_installments,payment_value
count,103886.000000,103886.000000,103886.000000
mean,1.092679,2.853349,154.100380
std,0.706584,2.687051,217.494064
min,1.000000,0.000000,0.000000
25%,1.000000,1.000000,56.790000
50%,1.000000,1.000000,100.000000
75%,1.000000,4.000000,171.837500
max,29.000000,24.000000,13664.080000



Itens Nulos
        order_id  payment_sequential  payment_type  payment_installments  \
0          False               False         False                 False   
1          False               False         False                 False   
2          False               False         False                 False   
3          False               False         False                 False   
4          False               False         False                 False   
...          ...                 ...           ...                   ...   
103881     False               False         False                 False   
103882     False               False         False                 False   
103883     False               False         False                 False   
103884     False               False         False                 False   
103885     False               False         False                 False   

        payment_value  
0               False  
1               False  
2 

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00



TAIL


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
99436,9c5dedf39a927c1b2549525ed64a053c,39bd1228ee8140590ac3aca26f2dfe00,delivered,2017-03-09 09:54:05,2017-03-09 09:54:05,2017-03-10 11:18:03,2017-03-17 15:08:01,2017-03-28 00:00:00
99437,63943bddc261676b46f01ca7ac2f7bd8,1fca14ff2861355f6e5f14306ff977a7,delivered,2018-02-06 12:58:58,2018-02-06 13:10:37,2018-02-07 23:22:42,2018-02-28 17:37:56,2018-03-02 00:00:00
99438,83c1379a015df1e13d02aae0204711ab,1aa71eb042121263aafbe80c1b562c9c,delivered,2017-08-27 14:46:43,2017-08-27 15:04:16,2017-08-28 20:52:26,2017-09-21 11:24:17,2017-09-27 00:00:00
99439,11c177c8e97725db2631073c19f07b62,b331b74b18dc79bcdf6532d51e1637c1,delivered,2018-01-08 21:28:27,2018-01-08 21:36:21,2018-01-12 15:35:03,2018-01-25 23:32:54,2018-02-15 00:00:00
99440,66dea50a8b16d9b4dee7af250b4be1a5,edb027a75a1449115f6b43211ae02a24,delivered,2018-03-08 20:57:30,2018-03-09 11:20:28,2018-03-09 22:11:59,2018-03-16 13:08:30,2018-04-03 00:00:00



INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB

DESCRIBE


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
count,99441,99441,99441,99441,99281,97658,96476,99441
unique,99441,99441,8,98875,90733,81018,95664,459
top,66dea50a8b16d9b4dee7af250b4be1a5,edb027a75a1449115f6b43211ae02a24,delivered,2018-08-02 12:06:07,2018-02-27 04:31:10,2018-05-09 15:48:00,2018-05-14 20:02:44,2017-12-20 00:00:00
freq,1,1,96478,3,9,47,3,522



Itens Nulos
       order_id  customer_id  order_status  order_purchase_timestamp  \
0         False        False         False                     False   
1         False        False         False                     False   
2         False        False         False                     False   
3         False        False         False                     False   
4         False        False         False                     False   
...         ...          ...           ...                       ...   
99436     False        False         False                     False   
99437     False        False         False                     False   
99438     False        False         False                     False   
99439     False        False         False                     False   
99440     False        False         False                     False   

       order_approved_at  order_delivered_carrier_date  \
0                  False                         False   
1     

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0



TAIL


,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
32946,a0b7d5a992ccda646f2d34e418fff5a0,moveis_decoracao,45.0,67.0,2.0,12300.0,40.0,40.0,40.0
32947,bf4538d88321d0fd4412a93c974510e6,construcao_ferramentas_iluminacao,41.0,971.0,1.0,1700.0,16.0,19.0,16.0
32948,9a7c6041fa9592d9d9ef6cfe62a71f8c,cama_mesa_banho,50.0,799.0,1.0,1400.0,27.0,7.0,27.0
32949,83808703fc0706a22e264b9d75f04a2e,informatica_acessorios,60.0,156.0,2.0,700.0,31.0,13.0,20.0
32950,106392145fca363410d287a815be6de4,cama_mesa_banho,58.0,309.0,1.0,2083.0,12.0,2.0,7.0



INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB

DESCRIBE


,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
count,32341.000000,32341.000000,32341.000000,32949.000000,32949.000000,32949.000000,32949.000000
mean,48.476949,771.495285,2.188986,2276.472488,30.815078,16.937661,23.196728
std,10.245741,635.115225,1.736766,4282.038731,16.914458,13.637554,12.079047
min,5.000000,4.000000,1.000000,0.000000,7.000000,2.000000,6.000000
25%,42.000000,339.000000,1.000000,300.000000,18.000000,8.000000,15.000000
50%,51.000000,595.000000,1.000000,700.000000,25.000000,13.000000,20.000000
75%,57.000000,972.000000,3.000000,1900.000000,38.000000,21.000000,30.000000
max,76.000000,3992.000000,20.000000,40425.000000,105.000000,105.000000,118.000000



Itens Nulos
       product_id  product_category_name  product_name_lenght  \
0           False                  False                False   
1           False                  False                False   
2           False                  False                False   
3           False                  False                False   
4           False                  False                False   
...           ...                    ...                  ...   
32946       False                  False                False   
32947       False                  False                False   
32948       False                  False                False   
32949       False                  False                False   
32950       False                  False                False   

       product_description_lenght  product_photos_qty  product_weight_g  \
0                           False               False             False   
1                           False               False   

,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto
3,cama_mesa_banho,bed_bath_table
4,moveis_decoracao,furniture_decor



TAIL


,product_category_name,product_category_name_english
66,flores,flowers
67,artes_e_artesanato,arts_and_craftmanship
68,fraldas_higiene,diapers_and_hygiene
69,fashion_roupa_infanto_juvenil,fashion_childrens_clothes
70,seguros_e_servicos,security_and_services



INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   product_category_name          71 non-null     object
 1   product_category_name_english  71 non-null     object
dtypes: object(2)
memory usage: 1.2+ KB

DESCRIBE


,product_category_name,product_category_name_english
count,71,71
unique,71,71
top,beleza_saude,health_beauty
freq,1,1



Itens Nulos
    product_category_name  product_category_name_english
0                   False                          False
1                   False                          False
2                   False                          False
3                   False                          False
4                   False                          False
..                    ...                            ...
66                  False                          False
67                  False                          False
68                  False                          False
69                  False                          False
70                  False                          False

[71 rows x 2 columns]


TABELA: vendedores


HEAD


,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP



TAIL


,seller_id,seller_zip_code_prefix,seller_city,seller_state
3090,98dddbc4601dd4443ca174359b237166,87111,sarandi,PR
3091,f8201cab383e484733266d1906e2fdfa,88137,palhoca,SC
3092,74871d19219c7d518d0090283e03c137,4650,sao paulo,SP
3093,e603cf3fec55f8697c9059638d6c8eb5,96080,pelotas,RS
3094,9e25199f6ef7e7c347120ff175652c3b,12051,taubate,SP



INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   int64 
 2   seller_city             3095 non-null   object
 3   seller_state            3095 non-null   object
dtypes: int64(1), object(3)
memory usage: 96.8+ KB

DESCRIBE


,seller_zip_code_prefix
count,3095.000000
mean,32291.059451
std,32713.453830
min,1001.000000
25%,7093.500000
50%,14940.000000
75%,64552.500000
max,99730.000000



Itens Nulos
      seller_id  seller_zip_code_prefix  seller_city  seller_state
0         False                   False        False         False
1         False                   False        False         False
2         False                   False        False         False
3         False                   False        False         False
4         False                   False        False         False
...         ...                     ...          ...           ...
3090      False                   False        False         False
3091      False                   False        False         False
3092      False                   False        False         False
3093      False                   False        False         False
3094      False                   False        False         False

[3095 rows x 4 columns]


In [ ]:
df_final = df_clientes.merge(
    df_pedidos, on='customer_id',
).merge(
    df_itens_pedidos, on='order_id',
).merge(
    df_produtos, on='product_id',
).merge(
    df_avaliacoes, on='order_id',
).merge(
    df_pagamentos, on='order_id',
).merge(
    df_vendedores, on='seller_id',
)

In [ ]:
print(df_final.shape)
df_final.head()

(117329, 39)


,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state,order_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,...,review_comment_message,review_creation_date,review_answer_timestamp,payment_sequential,payment_type,payment_installments,payment_value,seller_zip_code_prefix,seller_city,seller_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP,00e7ee1b050b8499577073aeb2a297a1,delivered,2017-05-16 15:05:35,2017-05-16 15:22:12,2017-05-23 10:47:57,...,NaN,2017-05-26 00:00:00,2017-05-30 22:34:40,1,credit_card,2,146.87,8577,itaquaquecetuba,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP,29150127e6685892b6eab3eec79f59c7,delivered,2018-01-12 20:48:24,2018-01-12 20:58:32,2018-01-15 17:14:59,...,NaN,2018-01-30 00:00:00,2018-02-10 22:43:29,1,credit_card,8,335.48,88303,itajai,SC
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP,b2059ed67ce144a36e2aa97d2c9e9ad2,delivered,2018-05-19 16:07:45,2018-05-20 16:19:10,2018-06-11 14:31:00,...,NaN,2018-06-15 00:00:00,2018-06-15 12:10:59,1,credit_card,7,157.73,8577,itaquaquecetuba,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP,951670f92359f4fe4a63112aa7306eba,delivered,2018-03-13 16:06:38,2018-03-13 17:29:19,2018-03-27 23:22:42,...,NaN,2018-03-29 00:00:00,2018-04-02 18:36:47,1,credit_card,1,173.30,8577,itaquaquecetuba,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP,6b7d50bd145f6fc7f33cebabd7e49d0f,delivered,2018-07-29 09:51:30,2018-07-29 10:10:09,2018-07-30 15:16:00,...,O baratheon è esxelente Amo adoro o baratheon,2018-08-10 00:00:00,2018-08-17 01:59:52,1,credit_card,8,252.25,14940,ibitinga,SP


### **Respostas às Questões de Carregamento e Entendimento dos Dados**

Com base nas execuções de código anteriores, as seguintes perguntas foram respondidas:

**Questões:**

*   **Quais as dimensões de cada tabela (quantidade de variáveis e informações)?**
    *   **Resposta:** Sim, o código imprimiu o número de linhas e colunas para cada DataFrame (`avaliacoes: 99224 linhas X 7 colunas`, etc.).

*   **Quais tabelas possuem textos (comentários, nomes de produtos entre outros)?**
    *   **Resposta:** Indiretamente sim, o método `.info()` executado para cada tabela lista as colunas com `dtype 'object'`, que geralmente correspondem a dados textuais.

*   **Quais tabelas possuem valores numéricos (preço, frete, parcelas)?**
    *   **Resposta:** Indiretamente sim, o método `.info()` para cada tabela lista as colunas com `dtype 'int64'` ou `'float64'`, que correspondem a dados numéricos.

*   **Quais limitações são identificadas dos dados (faltantes, repetidos ou colunas mal preenchidas)?**
    *   **Resposta:** Parcialmente sim, o método `.isnull()` para cada tabela permite identificar a presença de valores nulos. A identificação de dados repetidos ou colunas mal preenchidas requer verificações adicionais não realizadas explicitamente até o momento.
*   **Para que serve cada tabela dentro do fluxo de um e-commerce?**
    * **Resposta:**

**Perguntas sobre código:**

*   **Como analisar de forma geral as informações de cada tabela?**
    *   **Resposta:** Sim, a função `analise_df` e aplicada a todas as tabelas realiza uma análise geral, mostrando `head`, `tail`, `info`, `describe` e `isnull`.

*   **Como visualizar as primeiras, últimas linhas de uma tabela? E uma amostra?**
    *   **Resposta:** Sim, a função `analise_df` utiliza `display(df.head())` e `display(df.tail())` para visualizar as primeiras e últimas linhas, respectivamente.

*   **Como contar a quantidade total de valores nulos por coluna?**
    *   **Resposta:** Sim, a função `analise_df` executa `print(df.isnull())`, que mostra a presença de valores nulos em cada célula. Para a contagem total por coluna, seria necessário adicionar `.sum()` ao `df.isnull()`.

*   **Como verificar os tipos de dados de cada tabela?**
    *   **Resposta:** Sim, a função `analise_df` utiliza `df.info()` para exibir os tipos de dados de cada coluna, além da contagem de valores não nulos e uso de memória.

##### **Questões**


• Para que serve cada tabela dentro do fluxo de um e-commerce?

• Quais colunas funcionam como chaves primárias e quais funcionam como chaves estrangeiras que
conectam as tabelas entre si?

• Quais limitações são identificadas dos dados (faltantes, repetidos ou colunas mal preenchidas)?

• Alguma coluna precisa de descrição ou nome mais claro para o dicionário de dados?

##### **Perguntas sobre código**

• Como analisar de forma geral as informações de cada tabela?

• Como listar valores únicos de uma coluna específica?

• Como identificar rapidamente se existem linhas duplicadas?

• Como contar a quantidade total de valores nulos por coluna?

• Como verificar os tipos de dados de cada tabela?

• Como visualizar as primeiras, últimas linhas de uma tabela? E uma amostra?

Para visualizar as primeiras linhas de uma tabela, utilizamos .head(). Para fazer a visualização das últimas, vamos utilizar .tail().


##### **Perguntas de conclusão**

• Todas as tabelas foram carregadas corretamente? Alguma apresentou erro ou comportamento
inesperado?

• Os tipos de dados (datas, números, textos) foram reconhecidos corretamente pelo pandas?

• O que vocês perceberam de mais importante sobre a estrutura geral das tabelas?

• Quais relações entre as tabelas parecem essenciais para analisar um pedido completo?

• Que partes dos dados provavelmente precisarão de limpeza mais adiante?

• Que dificuldades você imagina ao juntar essas tabelas?

• As colunas de data estão no formato adequado para análise temporal?


# **2. Limpeza e Tratamentos**
Após o carregamento completo das nove tabelas, iniciaremos a etapa de limpeza e tratamento dos dados. Esta fase é essencial para garantir integridade, consistência e qualidade das informações antes das análises posteriores.

O objetivo aqui é identificar problemas estruturais, corrigir inconsistências e preparar os dados para operações como junções, agrupamentos, criação de métricas e análises estatísticas.

Nesta etapa realizaremos verificações e ajustes nas seguintes tabelas: **pedidos**, **itens_pedido**, **pagamentos**, **avaliacoes**, **clientes**, **vendedores**, **produtos**, **geolocalizacao** e na tabela **auxiliar**.

## **2.1 Principais problemas a verificar**

• **Valores nulos:** colunas de datas, preços, avaliações, coordenadas, nomes e códigos podem estar incompletos.

• **Dados duplicados:** especialmente nas tabelas **clientes**, **vendedores** e linhas repetidas em **geolocalização**.

• **Tipos incorretos:** datas lidas como texto, códigos numéricos armazenados como strings, colunas numéricas com separador inadequado.

• **Padronização de texto:** nomes de produtos com letras maiúsculas/minúsculas inconsistentes, categorias divergentes, comentários com símbolos especiais.

• **Inconsistências entre tabelas:** chaves secundárias sem correspondência (por exemplo, um id_pedido existente em **itens_pedido** mas ausente na tabela **pedidos**).

• **Registros inválidos:** datas invertidas, pagamentos com valores negativos ou fretes fora da realidade.

#### Tipos Incorretos

In [ ]:
#criando cópia dos dataframes para ajuste da tipagem dos dados
df_pedidos_type = df_pedidos.copy()
df_produtos_type = df_produtos.copy()
df_clientes_type = df_clientes.copy()
df_avaliacoes_type = df_avaliacoes.copy()
df_geolocalizacao_type = df_geolocalizacao.copy()
df_itens_pedidos_type = df_itens_pedidos.copy()
df_pagamentos_type = df_pagamentos.copy()
df_tabela_auxiliar_type = df_tabela_auxiliar.copy()
df_vendedores_type = df_vendedores.copy()

In [ ]:
#tipos dos dados antes dos ajustes (se necessários)
df_clientes_type.info()
print('-' * 50)
df_avaliacoes_type.info()
print('-' * 50)
df_geolocalizacao_type.info()
print('-' * 50)
df_itens_pedidos_type.info()
print('-' * 50)
df_pagamentos_type.info()
print('-' * 50)
df_tabela_auxiliar_type.info()
print('-' * 50)
df_vendedores_type.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB
--------------------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                99224 non-null  object
 1   order_id                 99224 non-null  object
 2   review_score             99224 non-null  int64 
 3   review_comment_title     11568 non-

In [ ]:
#ajustando tipagem dos dados incorretos
#PEDIDOS
df_pedidos_type['order_id'] = df_pedidos_type['order_id'].astype('string')
df_pedidos_type['customer_id'] = df_pedidos_type['customer_id'].astype('string')
df_pedidos_type['order_status'] = df_pedidos_type['order_status'].astype('category')
df_pedidos_type['order_purchase_timestamp'] = pd.to_datetime(df_pedidos_type['order_purchase_timestamp'], errors='coerce')
df_pedidos_type['order_approved_at'] = pd.to_datetime(df_pedidos_type['order_approved_at'], errors='coerce')
df_pedidos_type['order_delivered_carrier_date'] = pd.to_datetime(df_pedidos_type['order_delivered_carrier_date'], errors='coerce')
df_pedidos_type['order_delivered_customer_date'] = pd.to_datetime(df_pedidos_type['order_delivered_customer_date'], errors='coerce')
df_pedidos_type['order_estimated_delivery_date'] = pd.to_datetime(df_pedidos_type['order_estimated_delivery_date'], errors='coerce')

#CLIENTES
df_clientes_type['customer_id'] = df_clientes_type['customer_id'].astype('string')
df_clientes_type['customer_unique_id'] = df_clientes_type['customer_unique_id'].astype('string')
df_clientes_type['customer_city'] = df_clientes_type['customer_city'].astype('category')
df_clientes_type['customer_state'] = df_clientes_type['customer_state'].astype('category')

#AVALIAÇÕES
df_avaliacoes_type['review_id'] = df_avaliacoes_type['review_id'].astype('string')
df_avaliacoes_type['order_id'] = df_avaliacoes_type['order_id'].astype('string')
df_avaliacoes_type['review_comment_title'] = df_avaliacoes_type['review_comment_title'].astype('string')
df_avaliacoes_type['review_comment_message'] = df_avaliacoes_type['review_comment_message'].astype('string')
df_avaliacoes_type['review_creation_date'] = pd.to_datetime(df_avaliacoes_type['review_creation_date'], errors='coerce')
df_avaliacoes_type['review_answer_timestamp'] = pd.to_datetime(df_avaliacoes_type['review_answer_timestamp'], errors='coerce')

#GEOLOCALIZAÇÃO
df_geolocalizacao_type['geolocation_city'] = df_geolocalizacao_type['geolocation_city'].astype('category')
df_geolocalizacao_type['geolocation_state'] = df_geolocalizacao_type['geolocation_state'].astype('category')

#ITENS PEDIDOS
df_itens_pedidos_type['order_id'] = df_itens_pedidos_type['order_id'].astype('string')
df_itens_pedidos_type['product_id'] = df_itens_pedidos_type['product_id'].astype('string')
df_itens_pedidos_type['seller_id'] = df_itens_pedidos_type['seller_id'].astype('string')
df_itens_pedidos_type['shipping_limit_date'] = pd.to_datetime(df_itens_pedidos_type['shipping_limit_date'], errors='coerce')

#PAGAMENTOS
df_pagamentos_type['order_id'] = df_pagamentos_type['order_id'].astype('string')
df_pagamentos_type['payment_type'] = df_pagamentos_type['payment_type'].astype('category')

#TABELA AUXILIAR
df_tabela_auxiliar_type['product_category_name'] = df_tabela_auxiliar_type['product_category_name'].astype('string')
df_tabela_auxiliar_type['product_category_name_english'] = df_tabela_auxiliar_type['product_category_name_english'].astype('string')

#VENDEDORES
df_vendedores_type['seller_id'] = df_vendedores_type['seller_id'].astype('string')
df_vendedores_type['seller_city'] = df_vendedores_type['seller_city'].astype('category')
df_vendedores_type['seller_state'] = df_vendedores_type['seller_state'].astype('category')

In [ ]:
#tipos dos dados depois dos ajustes necessários
df_clientes_type.info()
print('-' * 50)
df_avaliacoes_type.info()
print('-' * 50)
df_geolocalizacao_type.info()
print('-' * 50)
df_itens_pedidos_type.info()
print('-' * 50)
df_pagamentos_type.info()
print('-' * 50)
df_tabela_auxiliar_type.info()
print('-' * 50)
df_vendedores_type.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype   
---  ------                    --------------  -----   
 0   customer_id               99441 non-null  string  
 1   customer_unique_id        99441 non-null  string  
 2   customer_zip_code_prefix  99441 non-null  int64   
 3   customer_city             99441 non-null  category
 4   customer_state            99441 non-null  category
dtypes: category(2), int64(1), string(2)
memory usage: 2.7 MB
--------------------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   review_id                99224 non-null  string        
 1   order_id                 99224 non-null  string        
 2   review_score             99224

In [ ]:
# Resumo Detalhado das Tabelas (incluindo Duplicatas por ID)

resumo = []

for nome, df in tabelas.items():
    # 1. Identifica colunas que provavelmente são IDs (contêm 'id' ou 'customer', 'order', etc)
    cols_id = [c for c in df.columns if 'id' in c.lower() or 'pk' in c.lower() or c.lower().startswith('customer') or c.lower().startswith('order') or c.lower().startswith('seller') or c.lower().startswith('product')]

    # 2. Calcula duplicatas na chave principal (se existir)
    dup_chave = 0
    chave_usada = "N/A"
    if cols_id:
        # Prioriza IDs que parecem ser chave primária da própria tabela
        pk_candidates = [c for c in cols_id if nome[:-1] + '_id' in c.lower() or nome + '_id' in c.lower() or c.lower() == 'review_id' or c.lower() == 'order_id' or c.lower() == 'customer_id' or c.lower() == 'seller_id' or c.lower() == 'product_id']

        if pk_candidates:
            chave_usada = pk_candidates[0]
            dup_chave = df.duplicated(subset=[chave_usada]).sum()
        elif cols_id:
            chave_usada = cols_id[0] # Usa o primeiro ID encontrado como fallback
            dup_chave = df.duplicated(subset=[chave_usada]).sum()

    resumo.append({
        "Tabela":       nome,
        "Linhas":       df.shape[0],
        "Colunas":      df.shape[1],
        "Nulos Total":  df.isnull().sum().sum(),
        "Nulos %":      f"{(df.isnull().sum().sum() / (df.size) * 100):.1f}%",
        "Dup. Linha":   df.duplicated().sum(),
        "Dup. ID":      dup_chave,
        "Chave Usada":  chave_usada
    })

df_resumo = pd.DataFrame(resumo)
print("Resumo detalhado das tabelas:")
display(df_resumo)

In [ ]:
# Tratamento e Combinação de Texto de Avaliações
original_total_reviews = len(df_avaliacoes)

def combine_review_text(row):
    title = row['review_comment_title']
    message = row['review_comment_message']

    # Convert to string and handle NaN as empty string
    title_str = str(title) if pd.notna(title) else ''
    message_str = str(message) if pd.notna(message) else ''

    if title_str and message_str:
        return f"{title_str} - {message_str}"
    elif title_str:
        return title_str
    elif message_str:
        return message_str
    else:
        return ''

df_avaliacoes['review_text_combined'] = df_avaliacoes.apply(combine_review_text, axis=1)

# Filter df_avaliacoes to remove rows where 'review_text_combined' is empty
df_avaliacoes = df_avaliacoes[df_avaliacoes['review_text_combined'] != ''].copy()

print(f"Total de avaliações originais: {original_total_reviews:,}")
print(f"Avaliações com texto combinado (título ou mensagem): {len(df_avaliacoes):,}")
print(f"Avaliações removidas (sem título e sem comentário original): {original_total_reviews - len(df_avaliacoes):,}")

In [ ]:
# Padronização de Texto em Avaliações (Minúsculas e Remoção de Acentos)

def strip_accents(text):
   """Remove accents from a string."""
   # Ensure text is a string before processing
   if isinstance(text, str):
       text = unicodedata.normalize('NFKD', text).encode('ascii', 'ignore').decode('utf-8')
   return text

df_avaliacoes['review_text_combined'] = df_avaliacoes['review_text_combined'].str.lower().apply(strip_accents)

print("Coluna 'review_text_combined' padronizada (minúsculas e sem acentos).")

In [ ]:
# Remoção de Duplicatas em `df_geolocalizacao`
antes = len(df_geolocalizacao)
df_geolocalizacao = df_geolocalizacao.drop_duplicates()
depois = len(df_geolocalizacao)

print(f"Duplicatas removidas em df_geolocalizacao: {antes - depois:,}")